src/main/java/com/minecolonies/coremod/tileentities/TileEntityRack.java

```diff
@@ -0,0 +1,502 @@
+package com.minecolonies.coremod.tileentities;
+
+import com.minecolonies.api.util.BlockPosUtil;
+import com.minecolonies.api.util.ItemStackUtils;
+import com.minecolonies.api.util.constant.Constants;
+import com.minecolonies.blockout.Log;
+import com.minecolonies.coremod.blocks.BlockRack;
+import com.minecolonies.coremod.entity.ai.item.handling.ItemStorage;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.init.Blocks;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.network.NetworkManager;
+import net.minecraft.network.play.server.SPacketUpdateTileEntity;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.items.CapabilityItemHandler;
+import net.minecraftforge.items.IItemHandlerModifiable;
+import net.minecraftforge.items.ItemStackHandler;
+import net.minecraftforge.items.wrapper.CombinedInvWrapper;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Predicate;
+
+/**
+ * Tile entity for the warehouse shelves.
+ */
+public class TileEntityRack extends TileEntity
+{
+    /**
+     * Tag used to store the neighbor pos to NBT.
+     */
+    private static final String TAG_NEIGHBOR = "neighbor";
+
+    /**
+     * Tag used to store the inventory to nbt.
+     */
+    private static final String TAG_INVENTORY = "inventory";
+
+    /**
+     * Tag used to store if the entity is the main.
+     */
+    private static final String TAG_MAIN = "main";
+
+    /**
+     * Tag compound of forge.
+     */
+    private static final int TAG_COMPOUND = 10;
+
+    /**
+     * Variable which determines if it is a single or doublechest.
+     */
+    private boolean single = true;
+
+    /**
+     * Neighbor position of the rack (double chest).
+     */
+    private BlockPos neighbor = BlockPos.ORIGIN;
+
+    /**
+     * Is this the main chest of the doubleChest.
+     */
+    private boolean isMain = false;
+
+    /**
+     * The content of the chest.
+     */
+    private final Map<ItemStorage, Integer> content = new HashMap<>();
```

Comment:
Use a BiMap here, allows for faster retrieval of the Slot for a given ItemStorage. Makes both the lookup for a Slot for a given ItemStorage as wel as a ItemStorage for a given Slot a lot faster.